klog_test.go

```diff
@@ -639,3 +640,299 @@ func TestInitFlags(t *testing.T) {
+
+	store *testLogger
+}
+
+func (l testLogger) Info(msg string, kvs ...interface{}) {
+	l.store.message = msg
+	l.store.keysAndValues = append(l.keysAndValues, kvs...)
+}
+
+func (l testLogger) Enabled() bool {
+	return l.enabled
+}
+
+func (l testLogger) Error(err error, msg string, kvs ...interface{}) {
+	l.store.err = err
+	l.store.message = msg
+	l.store.keysAndValues = append(l.keysAndValues, kvs...)
+}
+
+func (l testLogger) V(level int) logr.InfoLogger {
+	ril := testInfoLogger{}
+	ril.verbosity = level
+	ril.message = l.message
+	ril.keysAndValues = l.keysAndValues
+	ril.enabled = true
+	return ril
+}
+
+func (l testLogger) WithValues(kvs ...interface{}) logr.Logger {
+	rl := l
+	rl.keysAndValues = append(rl.keysAndValues, kvs...)
+	return rl
+}
+
+func (l testLogger) WithName(name string) logr.Logger {
+	rl := l
+	rl.name = append(rl.name, name)
+	return rl
+}
+
+func compareKeysAndValues(a, b []interface{}) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	for i := 0; i < len(a); i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+/*
+output method of loggingT is called at
+ * println method of loggingT
+ * printDepth method of loggingT
+ * printf method of loggingT
+ * printWithFileLine method of loggingT
+ * print method of loggingT (This method calls printDepth. So indirectly call output.)
+
+And above methods are called at
+ * Write method of logBridge
+ * Info method of Verbose
+ * Infoln method of Verbose
+ * Infof method of Verbose
+ * Info function
+ * InfoDepth function
+ * Infoln function
+ * Infof function
+ * Warning function
+ * WarningDepth function
+ * Warningln function
+ * Warningf function
+ * Error function
+ * ErrorDepth function
+ * Errorln function
+ * Errorf function
+ * Fatal function
+ * FatalDepth function
+ * Fatalln function
+ * Fatalf function
+ * Exit function
+ * ExitDepth function
+ * Exitln function
+ * Exitf function
+and
+  * receiver of output method is global variable logging in all these methods
+  * argument log of output method is
+     1. Verbose.logr at Info, Infoln and Infof method of Verbose with argument s as infoLog, or
+     2. global variable logging at all other these methods with argument s as any severity.
+
+So currently, there should be 4 situations in which logr.InfoLogger/logr.Logger is handled:
+ 1. log argument is set with severity as infoLog & logr field of receiver l is set
+     => Info method of l.logr is called.
+ 2. log argument is set with severity as infoLog & logr field of receiver l is not set
+     => Info method of log argument (type logr.InfoLogger) is called.
+ 3. logr field of receiver l is set & l.logr is passed as log argument with severity as any value
+     => * (3-1) If severity is errorLog, Error method of l.logr is called
+        * (3-2) If severity is not infoLog, Info method of l.logr is called
+
+This test function checks whether above 4 situations hold.
+*/
+func TestOutput(t *testing.T) {
+	// Situation 1
+	//
```

Comment:
different test cases should be defined in a table and be run as sub-tests. see "Table-driven tests using subtests" in https://blog.golang.org/subtests.

Reply:
^ this is ideally, if this will complicate your code too much you can break the cases in separate Test* functions.